display/download.go

```diff
@@ -83,56 +86,103 @@ func dlInit() {
 	})
 }
 
+func getMediaHandler(resp *gemini.Response) config.MediaHandler {
+	def := config.MediaHandler{
+		Cmd:      nil,
+		NoPrompt: false,
+	}
+
+	mediatype, _, err := mime.ParseMediaType(resp.Meta)
+	if err != nil {
+		return def
+	}
+
+	if ret, ok := config.MediaHandlers[mediatype]; ok {
+		return ret
+	}
+
+	splitType := strings.Split(mediatype, "/")
+	if ret, ok := config.MediaHandlers[splitType[0]]; ok {
+		return ret
+	}
+
+	if ret, ok := config.MediaHandlers["*"]; ok {
+		return ret
+	}
+
+	return def
+}
+
 // dlChoice displays the download choice modal and acts on the user's choice.
 // It should run in a goroutine.
 func dlChoice(text, u string, resp *gemini.Response) {
 	defer resp.Body.Close()
 
-	parsed, err := url.Parse(u)
-	if err != nil {
-		Error("URL Error", err.Error())
-		return
-	}
+	mediaHandler := getMediaHandler(resp)
+	var choice string
 
-	dlChoiceModal.SetText(text)
-	tabPages.ShowPage("dlChoice")
-	tabPages.SendToFront("dlChoice")
-	App.SetFocus(dlChoiceModal)
-	App.Draw()
+	if mediaHandler.NoPrompt {
+		choice = "Open"
+	} else {
+		dlChoiceModal.SetText(text)
+		tabPages.ShowPage("dlChoice")
+		tabPages.SendToFront("dlChoice")
+		App.SetFocus(dlChoiceModal)
+		App.Draw()
+		choice = <-dlChoiceCh
+	}
 
-	choice := <-dlChoiceCh
 	if choice == "Download" {
 		tabPages.HidePage("dlChoice")
 		App.Draw()
-		downloadURL(u, resp)
+		downloadURL(config.DownloadsDir, u, resp)
 		return
 	}
-	if choice == "Open in portal" {
-		// Open in mozz's proxy
-		portalURL := u
-		if parsed.RawQuery != "" {
-			// Remove query and add encoded version on the end
-			query := parsed.RawQuery
-			parsed.RawQuery = ""
-			portalURL = parsed.String() + "%3F" + query
-		}
-		portalURL = strings.TrimPrefix(portalURL, "gemini://") + "?raw=1"
-		ok := handleHTTP("https://portal.mozz.us/gemini/"+portalURL, false)
-		if ok {
-			tabPages.SwitchToPage(strconv.Itoa(curTab))
-			App.SetFocus(tabs[curTab].view)
-			App.Draw()
-		}
+	if choice == "Open" {
+		tabPages.HidePage("dlChoice")
+		App.Draw()
+		downloadAndOpen(u, resp)
 		return
 	}
 	tabPages.SwitchToPage(strconv.Itoa(curTab))
 	App.SetFocus(tabs[curTab].view)
 	App.Draw()
 }
 
+// downloadAndOpen performs the same actions as downloadURL except it also opens the file.
+// If there is no system viewer configured for the particular mediatype, it opens it
+// with the default system viewer.
+func downloadAndOpen(u string, resp *gemini.Response) {
```

Comment:
Can you rename this to just `open` or something? Because in the future this function might not download, but stream/pipe instead.

Reply:
Will do